File: DevelopArchitecture/DevelopArchitecture/architecture/baseVC/JBaseTableViewVC.m
//
//  JBaseTableViewVC.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 15/7/15.
//  Copyright (c) 2015年 Jeffrey. All rights reserved.
//

#import "JBaseTableViewVC.h"

@interface JBaseTableViewVC ()
@property(nonatomic,copy)reloadToSolveDataCallback m_reloadTableViewCallback;
@end

@implementation JBaseTableViewVC
@synthesize m_defaultImageView;
-(id)init{
    if(self=[super init]){
        self.m_isSearchTableViewList=NO;
         _m_isCanDeleteTableItem=NO;
    }
    return self;
}
-(void)setM_isSearchTableViewList:(BOOL)isSearchTableViewList{
    UIScrollView *tableview = [self funj_getTabeleView];
    tableview.m_currentPageType = isSearchTableViewList;
    _m_isSearchTableViewList = isSearchTableViewList;
}
maddProperyValue(m_dataArr, NSMutableArray)
maddProperyValue(m_searchDataArr, NSMutableArray)
maddProperyValue(m_dataSet, NSMutableSet)
maddProperyValue(m_searchDataSet, NSMutableSet)
maddProperyValue(m_multTypeDataModel, JTableDataArrModel)
maddProperyValue(m_searchMultTypeDataModel, JTableDataArrModel)
maddProperyValue(m_saveQuestionHeightDic, NSMutableDictionary)

-(UIView*)m_topView{
    if(!_m_topView){
        _m_topView =[UIView funj_getView:CGRectMake(0, 0, KWidth, 0) :COLOR_WHITE_DARK];
        [self.view addSubview: _m_topView];
        _m_topTableViewLine =[UIImageView funj_getLineImageView:CGRectMake(0, 0, KWidth, 1)];
        [_m_topView addSubview:_m_topTableViewLine];
     }
    return _m_topView;
}
-(JSearchBar*)m_searchBar{
    if(!_m_searchBar){
        _m_searchBar = [[JSearchBar alloc] initWithFrame:CGRectMake(0,0,KWidth,37)];
        _m_searchBar.searchDelegate = self;
        _m_searchBar.m_filletValue = JFilletMake(0.5, 37/2, COLOR_LINE_GRAY_DARK);
        _m_searchBar.hidden = YES;
        [_m_searchBar funj_reloadSearchState:YES :YES];
        [_m_topView addSubview:_m_searchBar];
        
        _m_blackImageView=[UIImageView funj_getBlackAlphaView:CGRectMake(0, 0, KWidth, KHeight)];
        _m_blackImageView.hidden=YES;
        [self.view addSubview:_m_blackImageView];
        __weak typeof(self) weakSelf = self;
        [_m_blackImageView  funj_whenTouchedDown:^(UIView *sender) {
            __strong typeof(weakSelf) strongSelf = weakSelf;
            [strongSelf.m_searchBar endEditing:YES];
            [strongSelf funj_searchBarState:NO];
            strongSelf.m_isSearchTableViewList=NO;
            [strongSelf funj_reloadData];
        }];
        [self.view addSubview:_m_blackImageView];
    }
    return _m_searchBar;
}
-(UITableView*)m_tableView{
    if(!_m_tableView){
        _m_tableView = kcreateTableViewWithDelegate(self);
    }
    return _m_tableView ;
}
-(UIScrollView*)funj_getTabeleView{
    UIScrollView *tableview = nil;
    if([self isKindOfClass:NSClassFromString(@"JBaseCollectionVC")]){
        tableview = [self valueForKeyPath:@"m_collectionView"];
    }else{
        tableview = [self valueForKeyPath:@"m_tableView"];
    }
    return tableview;
}
-(void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    UITableView* tableview = (UITableView*)[self funj_getTabeleView];
    self.m_defaultImageView.hidden = tableview.visibleCells.count>0;
}

-(void)viewDidLoad{
    [super viewDidLoad];
    [self addNoDataDefaultView];
    if(![self isKindOfClass:NSClassFromString(@"JBaseCollectionVC")]){
        [self.view addSubview:self.m_tableView];
    }
 }

-(void)addNoDataDefaultView{
     m_defaultImageView=[UIImageView funj_getImageView:CGRectZero image:@"uu_tableview_default_icon"];
    
    UILabel *contentLabel =[UILabel funj_getLabel:CGRectMake(0, 0, 200, 20) :LocalStr(@"Here is a wilderness ...Nothing left") :JTextFCMakeAlign(PUBLIC_FONT_SIZE17,COLOR_TEXT_GRAY_DARK,NSTextAlignmentCenter)];
    [m_defaultImageView addSubview:contentLabel];
    contentLabel.tag = 9993;
}
-(void)funj_reloadBaseViewParameter:(CGRect)m_topViewFrame :(CGRect)tableViewFrame :(BOOL)searchBarHidden{
    if(!CGRectEqualToRect(m_topViewFrame, CGRectZero)){
         self.m_topView.frame=m_topViewFrame;
         self.m_topTableViewLine.frame=CGRectMake(_m_topTableViewLine.left, _m_topView.height-1, _m_topView.width, _m_topTableViewLine.height);
    }
    UIScrollView *tableview = [self funj_getTabeleView];
    tableview.frame=tableViewFrame;

    if(!searchBarHidden){
        self.m_searchBar.hidden = searchBarHidden;
        _m_searchBar.width = m_topViewFrame.size.width;
        CGPoint point = [self.view convertPoint:_m_searchBar.origin fromView:_m_searchBar];
        self.m_blackImageView.frame=CGRectMake(_m_blackImageView.left, point.y+_m_searchBar.height  , _m_topView.width, CGRectGetHeight(tableViewFrame)+(CGRectGetMinY(tableViewFrame)-_m_topTableViewLine.top));
    }
 
    CGFloat width = IS_IPAD ? 300 : 200;
    m_defaultImageView.frame=CGRectMake(CGRectGetWidth(tableViewFrame)/2-width/2, MAX(CGRectGetHeight(tableViewFrame)/2-width/2-60, 0), width, width);
    UILabel *contentLabel =[m_defaultImageView viewWithTag:9993];
    contentLabel.frame = CGRectMake((m_defaultImageView.width-tableview.width)/2, m_defaultImageView.height+10, tableview.width, 20);
    [tableview addSubview:m_defaultImageView];
}

-(void)funj_reloadTableViewToSolveData:(reloadToSolveDataCallback)callback{
    if(callback){
        self.m_reloadTableViewCallback=callback;
        __weak typeof(self) weakSelf = self;
        UIScrollView *tableview = [self funj_getTabeleView];
 
        [tableview funj_addFooterWithCallback:^(NSInteger page) {
            __strong typeof(weakSelf) strongSelf = weakSelf;
            strongSelf.m_isReloadNewDataing = NO;
            if(strongSelf. m_reloadTableViewCallback)strongSelf.m_reloadTableViewCallback(NO,strongSelf.m_isSearchTableViewList,page);
        }];
        [tableview funj_addHeaderWithCallback:^(NSInteger page) {
            __strong typeof(weakSelf) strongSelf = weakSelf;
            strongSelf.m_isReloadNewDataing = YES;
            if(strongSelf. m_reloadTableViewCallback)strongSelf.m_reloadTableViewCallback(YES,strongSelf.m_isSearchTableViewList,page);
        }];
    }
}
#pragma mark - TableView DataSource
-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    if(self.m_isSearchTableViewList)return [_m_searchDataArr count];
    else return [_m_dataArr count];
}
//-(CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath{
//    return 40;
//}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
//    NSDictionary *dic =[self.m_dataArr objectAtIndex:indexPath.row];
//    if([self.m_saveQuestionHeightDic objectForKey:dic[@"classroomId"]]){
//        return [self.m_saveQuestionHeightDic doubleWithKey:dic[@"classroomId"]];
//    }
    return 50;
}
//继承的此类的，要覆盖这个方法
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellIndentifier];
    return cell;
}
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
}
-(void)tableView:(UITableView *)tableView willDisplayCell:(nonnull UITableViewCell *)cell forRowAtIndexPath:(nonnull NSIndexPath *)indexPath{
    self.m_defaultImageView.hidden=YES;
}
-(void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section{
    self.m_defaultImageView.hidden=YES;
}
- (void)tableView:(UITableView *)tableView willDisplayFooterView:(UIView *)view forSection:(NSInteger)section{
    self.m_defaultImageView.hidden=YES;
}
//***************************删除cell***************************
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath{
    if(self.m_isCanDeleteTableItem) return YES;
    return NO;
}

-(UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath{
    UITableViewCellEditingStyle result = UITableViewCellEditingStyleNone;//默认没有编辑风格
    if ([tableView isEqual:_m_tableView]) {
        result = UITableViewCellEditingStyleDelete;//设置编辑风格为删除风格
    }
    return result;
}
-(void)setEditing:(BOOL)editing animated:(BOOL)animated{//设置是否显示一个可编辑视图的视图控制器。
    [super setEditing:editing animated:animated];
    [_m_tableView setEditing:editing animated:animated];//切换接收者的进入和退出编辑模式。
}
//设置删除标题
- (NSString *)tableView:(UITableView *)tableView titleForDeleteConfirmationButtonForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    return LocalStr(@"Delete");
}
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        [JAppViewTools funj_showAlertBlocks:self  :nil :LocalStr(@"Do you want to delete? ") :^(id strongSelf, NSInteger index) {
            
        }];
    }
}
-(void)funj_addCellCallbackHeight:(JBaseTableViewCell*)cell :(NSString*)idKey{
    __weak typeof(self) weakSelf = self;
     cell.m_callbackHeight = ^(NSString*idKey,CGFloat height) {
        dispatch_async(dispatch_get_global_queue(0, 0), ^{
            __strong typeof(weakSelf) strongSelf=weakSelf;
            if(strongSelf && [[strongSelf.m_saveQuestionHeightDic objectForKey:idKey] floatValue] != height){
                [strongSelf.m_saveQuestionHeightDic setObject:[NSString stringWithFormat:@"%f",height] forKey:idKey];
             }
        });
    };
}
#pragma mark JsearchBar delegate
- (BOOL)funj_searchShouldBeginEditing:(UITextField *)textField{
    if([textField.text length] <=0){
        self.m_isSearchTableViewList=NO;
        [self funj_searchBarState:YES];
    }else{
        self.m_isSearchTableViewList=YES;
        [self funj_searchBarState:NO];
    }
    return YES;
}
-(BOOL)funj_search:(UITextField *)textField inRange:(NSRange)range replacementString:(NSString *)string{
    if(range.length<=0 || (textField.text.length>1 && range.length>=1)){
        self.m_isSearchTableViewList = YES;
        [self funj_searchBarState:NO];
    }else{
        self.m_isSearchTableViewList = NO;
        [self funj_searchBarState:YES];
    }
    return YES;
}
-(void)funj_searchDidEndEditing:(UITextField *)textField{
    if(textField.text.length>0){
        self.m_isSearchTableViewList = YES;
    }else{
        self.m_isSearchTableViewList = NO;
    }
    [self funj_searchBarState:NO];
    [self funj_reloadData];
}
-(void)funj_searchReturnButtonClicked:(UITextField*)textField{
    self.m_isSearchTableViewList = YES;
    [self funj_searchBarState:NO];
}
-(void)funj_searchCancelButtonClicked:(UITextField *)textField{
    self.m_isSearchTableViewList = NO;
    [self funj_searchBarState:NO];
}

-(void)funj_searchBarState:(BOOL)isShowBlack{
     _m_blackImageView.hidden=!isShowBlack;
}
-(void)funj_solverToSetData:(NSArray*)data :(NSMutableSet*)targetArr{
    if(!targetArr){
        targetArr = self.m_isSearchTableViewList ? self.m_searchDataSet : self.m_dataSet;
    }
    if(self.m_isReloadNewDataing){
        [targetArr removeAllObjects];
        [self funj_reloadData];
    }
    self.m_isReloadNewDataing= NO;
    
    if(!data || [data count]<=0 ){
        self.m_defaultImageView.hidden = targetArr.count;
        return;
    }
    [targetArr addObjectsFromArray:data];
    self.m_defaultImageView.hidden = targetArr.count;
    [self funj_reloadData];
}
-(void)funj_solverToSubRepeatData:(NSArray*)data :(NSMutableArray*)targetArr :(NSString*)key{
    [self funj_solverToSubRepeatData:data :targetArr :key :nil];
}
-(void)funj_solverToSubRepeatData:(NSArray*)data :(NSMutableArray*)targetArr :(NSString*)key :(NSString*)model{
    //    if(!data || [data count]<=0 || !key || key.length<=0) return;
    if(!targetArr){
        targetArr = self.m_isSearchTableViewList ? self.m_searchDataArr : self.m_dataArr;
    }
    if(self.m_isReloadNewDataing){
        [targetArr removeAllObjects];
        [self funj_reloadData];
    }
    self.m_isReloadNewDataing= NO;
    
    if(!data || [data count]<=0 ){
        self.m_defaultImageView.hidden = targetArr.count;
        return;
    }
    NSMutableArray *array=[[NSMutableArray alloc]initWithArray:targetArr];
    
    if(model && model.length>0){
        SEL sel=NSSelectorFromString(key);
        
        [data enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            __block NSUInteger isHas = NSNotFound;
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            id object= [obj performSelector:sel];
            [array enumerateObjectsUsingBlock:^(id _Nonnull obj2, NSUInteger idx2, BOOL * _Nonnull stop2) {
                id object2= [obj2 performSelector:sel];
                if([[NSString stringWithFormat:@"%@",object] isEqualToString:[NSString stringWithFormat:@"%@",object2]]){
                    isHas = idx2; *stop2=YES;
                }
            }];
#pragma clang diagnostic pop
            if(isHas == NSNotFound && obj){
                [array addObject:obj];
            }else{
                if(isHas < array.count){
                    [array insertObject:obj atIndex:isHas];
                }else{
                    [array addObject:obj];
                }
            }
            [array removeObjectAtIndex:isHas+1];
            
        }];
        
    }else{
        [data enumerateObjectsUsingBlock:^(NSDictionary*  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            __block NSUInteger isHas = NSNotFound;
            [array enumerateObjectsUsingBlock:^(NSDictionary*  _Nonnull obj2, NSUInteger idx2, BOOL * _Nonnull stop2) {
                if([obj[key] isKindOfClass:[NSString class]] &&  [obj2[key] isEqualToString:obj[key]]){
                    isHas = idx2; *stop2=YES;
                }else if([obj[key] isKindOfClass:[NSNumber class]] && [obj[key] intValue] == [obj2[key] intValue]){
                    isHas = idx2; *stop2=YES;
                }
            }];
            
            if(isHas == NSNotFound ){
                [array addObject:obj];
            }else{
                if(isHas < array.count){
                    [array insertObject:obj atIndex:isHas];
                }else{
                    [array addObject:obj];
                }
                [array removeObjectAtIndex:isHas+1];
            }
        }];
    }
    
    if(targetArr.count >0){
        [targetArr removeAllObjects];
    }
    [targetArr addObjectsFromArray:array];
    self.m_defaultImageView.hidden = targetArr.count;
    [self funj_reloadData];
}
-(void )funj_reloadData{
    [self.m_tableView reloadData];
}
 
@end

File: DeveloperArchitecture/DeveloperArchitecture/architecture/baseVC/JBaseTableViewVC.swift
//
//  JBaseTableViewVC.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/10/30.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit

typealias kreloadToSolveDataCallback = ((_ isHead : Bool , _ type : String , _ page : Int) -> ())

let kCellIndentifier = "cellIndentifier"

class JBaseTableViewVC : JBaseViewController, JSearchBarDelegate, UITableViewDelegate,UITableViewDataSource {
    lazy var m_topView : UIView = {
        let topView = UIView(i: CGRect(x: 0, y: 0, width: kWidth, height: 0), bg: kColor_White_Dark)
        self.view.addSubview(topView)
        m_topTableViewLine = UIImageView(i_line: CGRect(x: 0, y: 0, width: kWidth, height: 1))
        topView.addSubview(m_topTableViewLine!)
        return topView
    }()
    var m_topTableViewLine : UIView?
    var m_blackImageView : UIImageView?
    var m_isCanDeleteTableItem : Bool  = false
    var m_isReloadNewDataing : Bool = false
    
    lazy var m_dataArr : [Any] = { return [] }()
    private lazy var m_saveQuestionHeightDic : [String : CGFloat] = { return [:] } ()
    
    var m_currentPageType : String = "default" { // 多种类型数据
        willSet {
            let tableView = self.funj_getTabeleView()
            tableView?.funj_setCurrentPageType(newValue)
        }
    }
    
    lazy var m_searchBar : JSearchBar = {
        let searchBar = JSearchBar(frame: CGRect(x: 0, y: 0, width: kWidth, height: 37))
        searchBar.m_searchDelegate = self
        searchBar.m_filletValue = JFilletValue(w: 0.5, r: 37/2, c: kColor_Line_Gray_Dark)
        searchBar.isHidden = true
        searchBar.funj_reloadSearchState(needIcon: true, needCancel: true)
        m_topView.addSubview(searchBar)
        
        m_blackImageView = UIImageView(i_blackAlpha: CGRect(x: 0, y: 0, width: kWidth, height: kHeight))
        m_blackImageView?.isHidden = true
        self.view.addSubview(m_blackImageView!)
        m_blackImageView?.funj_whenLongPressed({ (view) in
            self.m_searchBar.endEditing(true)
            self.funj_searchBarState(false)
        })
        self.view.addSubview(m_blackImageView!)
        return searchBar
    }()
    lazy var m_tableView : UITableView = {
        let tableView =  kcreateTableViewWithDelegate(self)
        return tableView
    }()
    lazy var m_defaultImageView : UIImageView = {
        let defaultImageView = UIImageView(i: CGRectZero, image: "uu_tableview_default_icon")
        let contentLabel = UILabel(i: CGRect(x: 0, y: 0, width: 200, height: 20), title: "Here is a wilderness ...Nothing left", textFC: JTextFC(f: kFont_Size17, c: kColor_Text_GRAY_Dark, a: .center))
        defaultImageView.addSubview(contentLabel)
        contentLabel.tag = 9993
        return defaultImageView
    }()
    var m_reloadTableViewCallback : kreloadToSolveDataCallback?
    
    func funj_getTabeleView() -> UIScrollView? {
        if self is JBaseCollectionVC {
            let collectionView =  self.value(forKeyPath: "m_collectionView") as? UIScrollView
            return collectionView
        } else {
            return m_tableView
        }
    }
}
extension JBaseTableViewVC {
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        m_defaultImageView.isHidden = self.m_tableView.visibleCells.count > 0
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        if self.isKind(of: JBaseCollectionVC.self) == false {
            self.view.addSubview(self.m_tableView)
        }
    }
}

extension JBaseTableViewVC {
    func funj_reloadData() {
        let tableView = self.funj_getTabeleView()
        (tableView as? UITableView)?.reloadData()
        (tableView as? UICollectionView)?.reloadData()
    }

    func funj_reloadTableView(_ topViewFrame : CGRect , table tableViewFrame : CGRect , isHidden searchBarHidden : Bool = true){
        if topViewFrame != CGRectZero {
            m_topView.frame = topViewFrame
            m_topTableViewLine?.frame = CGRect(x: m_topTableViewLine!.left, y: m_topView.height - 1, width: m_topView.width, height: m_topTableViewLine!.height)
        }
        let tableView = self.funj_getTabeleView()
        tableView?.frame = tableViewFrame
        
        if searchBarHidden == false {
            m_searchBar.isHidden = searchBarHidden
            m_searchBar.width = topViewFrame.width
            let point = self.view.convert(m_searchBar.origin, from: m_searchBar)
            m_blackImageView?.frame = CGRect(x: m_blackImageView!.left, y: point.y + m_searchBar.height, width: m_topView.width, height: tableViewFrame.height + (tableViewFrame.origin.y - (m_topTableViewLine?.top ?? 0)))
        }
        
        let width : CGFloat = kis_IPad ? 300.0 : 200.0 ;
        m_defaultImageView.frame = CGRect(x: (tableViewFrame.width - width)/2 , y: max(tableViewFrame.height/2 - width/2 - 60, 0), width: width, height: width)
        let contentLabel = m_defaultImageView.viewWithTag(9993)
        contentLabel?.frame = CGRect(x: (m_defaultImageView.width - (tableView?.width ?? 0))/2, y: m_defaultImageView.height + 10, width: tableView?.width ?? 0, height: 20)
        tableView?.addSubview(m_defaultImageView)
    }
    func funj_addLoadingCallback(_ callback : @escaping kreloadToSolveDataCallback){
        self.m_reloadTableViewCallback = callback
        let tableview = self.funj_getTabeleView()
        tableview?.funj_addHeader(callback: { [weak self] (type , page) in
            self?.m_isReloadNewDataing = true
            self?.m_reloadTableViewCallback?(true , type , page )
        })
        tableview?.funj_addFooter(callback: { [weak self] (type , page) in
            self?.m_isReloadNewDataing = false
            self?.m_reloadTableViewCallback?(false , type , page )
        })
    }
}

extension JBaseTableViewVC {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return m_dataArr.count
    }
    
//    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
//        return 40
//    }
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 100
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let tableviewCell = tableView.dequeueReusableCell(withIdentifier: kCellIndentifier)
        tableviewCell?.backgroundColor = krandomColor
        return tableviewCell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
    }
}
extension JBaseTableViewVC {
    func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return self.m_isCanDeleteTableItem
    }
    func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        var result : UITableViewCell.EditingStyle = .none
        if tableView == m_tableView {
            result = .delete
        }
        return result
    }
    override func setEditing(_ editing: Bool, animated: Bool) {
        super.setEditing(editing, animated: animated)
        m_tableView.setEditing(editing, animated: animated)
    }
    func tableView(_ tableView: UITableView, titleForDeleteConfirmationButtonForRowAt indexPath: IndexPath) -> String? {
        return kLocalStr("Delete")
    }
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            _ = JAppViewTools.funj_showAlertBlock(nil, message: kLocalStr("Do you want to delete? "), buttonArr: [kLocalStr("Cancel"),kLocalStr("Confirm")]) { (index) in
                
            }
        }
    }
}

extension JBaseTableViewVC {
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        m_defaultImageView.isHidden = true
    }
    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        m_defaultImageView.isHidden = true
    }
    func tableView(_ tableView: UITableView, willDisplayFooterView view: UIView, forSection section: Int) {
        m_defaultImageView.isHidden = true
    }
}

extension JBaseTableViewVC {
    func funj_addCellCallbackHeight(cell : JBaseTableViewCell , idKey : String) {
        cell.m_callback = { [weak self ](idKey, height) in
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                if let height1 = self?.m_saveQuestionHeightDic[idKey] , height1 != height {
                    self?.m_saveQuestionHeightDic[idKey] = height
                }
            }
        }
    }

    func funj_solverToSubRepeatData<T : JBaseDataModel>(data : [Any] , targeArr : inout [Any]? , key : String , model : T? = nil) {
        var array : [Any] = []
        if targeArr == nil {
            targeArr = m_dataArr
        }

        if m_isReloadNewDataing {
            targeArr?.removeAll()
            self.funj_reloadData()
        }
        
        for obj in targeArr! {
            array.append(obj)
        }
        
        if model == nil {
            for obj in data {
                var isHas = NSNotFound ; var index = 0
                for obj2 in array {
                    if let obj = obj as? [String : Any] , let value = obj[key] as? String {
                        if let obj2 = obj2 as? [String : Any] , let value2 = obj2[key] as? String , value == value2 {
                            isHas = index ; break
                        }
                    }
                    index += 1
                }
                if isHas == NSNotFound {
                    array.append(obj)
                } else {
                    if isHas < array.count {
                        array.insert(obj, at: isHas)
                    } else {
                        array.append(obj)
                    }
                    array.remove(at: isHas + 1)
                }
            }
        } else {
            for obj in data {
                var isHas = NSNotFound ; var index : Int = 0
                for obj2 in array {
                    
                    if let obj = obj as? T , let value = obj.value(forKey: key) {
                        if let obj2 = obj2 as? T , let value2 = obj2.value(forKey: key){
                            if let value = value as? String , let value2 = value2 as? String , value == value2 {
                                isHas = index ; break
                            } else {
//                                assert(1 != 1, " repeat data ")
//                                isHas = index ; break
                            }
                        }
                    }
                    index += 1
                }
                if isHas == NSNotFound {
                    array.append(obj)
                } else {
                    if isHas < array.count {
                        array.insert(obj, at: isHas)
                    } else {
                        array.append(obj)
                    }
                    array.remove(at: isHas + 1)
                }
            }
        }
        targeArr?.removeAll()
        for obj in array {
            targeArr?.append(obj)
        }
        m_defaultImageView.isHidden = targeArr?.count ?? 0 > 0
        self.funj_reloadData()
    }
}
//MARK JsearchBar delegate
extension JBaseTableViewVC {
    func funj_searchShouldBeginEditing(_ textField: UITextField) -> Bool {
        if textField.text?.count ?? 0 <= 0 {
            funj_searchBarState(true)
        } else {
            funj_searchBarState(false)
        }
        return true
    }
    func funj_search(_ textField: UITextField, range: NSRange, string: String) -> Bool {
        var isShow = true
        if range.length <= 0 { isShow = false }
        if range.length >= 1 {
            if textField.text?.count ?? 0 > 1 { isShow = false}
            else  if string.count > 0{
                if textField.text?.hasSuffix(string) ?? false == false{
                    isShow = false
                }
            }
        }
        funj_searchBarState(isShow)
        return true
    }
    func funj_searchDidEndEditing(_ textField: UITextField) {
        funj_searchBarState(false)
        funj_reloadData()
    }
    func funj_searchReturnButtonClicked(_ textField: UITextField) -> Bool {
        funj_searchBarState(false)
        self.view.endEditing(true)
        return true
    }
    func funj_searchCancelButtonClicked(_ textField: UITextField) {
        funj_searchBarState(false)
        self.view.endEditing(true)
    }
    func funj_searchBarState(_ isshowBlack : Bool) {
        m_blackImageView?.isHidden = !isshowBlack
    }
}

func kcreateTableViewWithDelegate(_ delegate : AnyObject) -> UITableView {
    let tableView = UITableView(frame: CGRect(x: 0, y: 0, width: kWidth, height: kHeight), style: .plain)
    tableView.tag = 939003
    tableView.backgroundColor = UIColor.clear
    tableView.rowHeight = 100
    tableView.showsVerticalScrollIndicator = false
    tableView.showsHorizontalScrollIndicator = false
    tableView.delegate = (delegate as? UITableViewDelegate)
    tableView.dataSource = (delegate as? UITableViewDataSource)
    tableView.separatorStyle = .none
    tableView.register(JBaseTableViewCell.self, forCellReuseIdentifier: kCellIndentifier)
    if #available(iOS 11.0, *) {
        tableView.contentInsetAdjustmentBehavior = .never
    }
    return tableView
}





